WalletWasabi.Fluent/Controls/CurrencyEntryBox/CurrencyEntryBox.axaml.cs

```diff
@@ -0,0 +1,375 @@
+		AvaloniaProperty.Register<CurrencyEntryBox, CurrencyInputViewModel?>(nameof(ViewModel));
+
+	public static readonly StyledProperty<CurrencyFormat> CurrencyFormatProperty =
+		AvaloniaProperty.Register<CurrencyEntryBox, CurrencyFormat>(nameof(CurrencyFormat), defaultValue: CurrencyFormat.Btc);
+
+	public static readonly StyledProperty<CurrencyValue> ValueProperty =
+		AvaloniaProperty.Register<CurrencyEntryBox, CurrencyValue>(nameof(Value), enableDataValidation: true);
+
+	// TODO: these would be better as attached properties of the Behavior
+	public static readonly StyledProperty<string?> ClipboardSuggestionProperty =
+		AvaloniaProperty.Register<CurrencyEntryBox, string?>(nameof(ClipboardSuggestion), defaultBindingMode: BindingMode.TwoWay);
+
+	public static readonly StyledProperty<ICommand> ApplySuggestionCommandProperty =
+		AvaloniaProperty.Register<CurrencyEntryBox, ICommand>(nameof(ApplySuggestionCommand));
+
+	private CompositeDisposable _disposables = new();
+	private bool _isUpdating;
+	private bool _isUpdatingSelection;
+
+	public CurrencyEntryBox()
+	{
+		SetCurrentValue(TextProperty, "");
+
+		AddHandler(KeyDownEvent, CustomOnKeyDown, RoutingStrategies.Tunnel);
+		AddHandler(PointerMovedEvent, CustomPointerMove, RoutingStrategies.Tunnel);
+		AddHandler(PointerPressedEvent, CustomPointerPressed, RoutingStrategies.Tunnel);
+
+		this.GetObservable(ViewModelProperty)
+			.Do(viewModel =>
+			{
+				_disposables.Dispose();
+				_disposables = new();
+
+				if (viewModel is not { })
+				{
+					return;
+				}
+
+				SetCurrentValue(CurrencyFormatProperty, viewModel.CurrencyFormat);
+
+				viewModel.WhenAnyValue(x => x.InsertPosition)
+						 .BindTo(this, x => x.CaretIndex)
+						 .DisposeWith(_disposables);
+
+				viewModel.WhenAnyValue(x => x.Text)
+						 .BindTo(this, x => x.Text)
+						 .DisposeWith(_disposables);
+
+				viewModel.WhenAnyValue(x => x.Value)
+						 .Do(v => SetCurrentValue(ValueProperty, v))
+						 .Subscribe()
+						 .DisposeWith(_disposables);
+
+				viewModel.WhenAnyValue(x => x.SelectionStart, x => x.SelectionEnd)
+						 .Where(_ => !_isUpdatingSelection)
+						 .Do(t =>
+						 {
+							 var (start, end) = t;
+
+							 if (start is null || end is null)
+							 {
+								 SelectionStart = CaretIndex;
+								 SelectionEnd = CaretIndex;
+							 }
+							 else
+							 {
+								 SelectionStart = start.Value;
+								 SelectionEnd = end.Value;
+							 }
+						 })
+						 .Subscribe()
+						 .DisposeWith(_disposables);
+
+				this.GetObservable(CaretIndexProperty)
+					.Do(viewModel.SetInsertPosition)
+					.Subscribe()
+					.DisposeWith(_disposables);
+
+				this.GetObservable(SelectionStartProperty)
+					.CombineLatest(this.GetObservable(SelectionEndProperty))
+					.Throttle(TimeSpan.FromMilliseconds(50), RxApp.MainThreadScheduler)
+					.Do(t =>
+					{
+						if (_isUpdating || t.First == t.Second || CurrencyFormat is null)
```

Comment:
I don't get the `CurrencyFormat` here. It has a default value in this class and on ViewModel level it can never be null. Why are we checking for null here?

Reply:
Since `CurrencyFormat` is bindable, it might be bound to a null value in other cases. I think keeping this extra check here doesn't hurt.